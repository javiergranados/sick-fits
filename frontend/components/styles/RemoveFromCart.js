import styled from 'styled-components';

const RemoveFromCart = styled.button`
  font-size: 3rem;
  background: none;
  border: 0;
  &:hover {
    color: ${({ theme }) => theme.red};
    cursor: pointer;
  }
`;

export { RemoveFromCart };
