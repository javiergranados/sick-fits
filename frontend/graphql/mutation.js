import gql from 'graphql-tag';

const CREATE_ITEM = gql`
  mutation createItem($title: String!, $description: String!, $price: Int!, $image: String, $largeImage: String) {
    createItem(title: $title, description: $description, price: $price, image: $image, largeImage: $largeImage) {
      id
    }
  }
`;

const UPDATE_ITEM = gql`
  mutation updateItem($id: ID!, $title: String, $description: String, $price: Int) {
    updateItem(id: $id, title: $title, description: $description, price: $price) {
      id
      title
      description
      price
    }
  }
`;

const DELETE_ITEM = gql`
  mutation deleteItem($id: ID!) {
    deleteItem(id: $id) {
      id
      title
    }
  }
`;

const SIGN_UP = gql`
  mutation signup($email: String!, $name: String!, $password: String!) {
    signup(email: $email, name: $name, password: $password) {
      id
      email
      name
    }
  }
`;

const SIGN_IN = gql`
  mutation signin($email: String!, $password: String!) {
    signin(email: $email, password: $password) {
      id
      email
      name
      password
    }
  }
`;

const SIGN_OUT = gql`
  mutation signout {
    signout {
      message
    }
  }
`;

const REQUEST_RESET = gql`
  mutation requestReset($email: String!) {
    requestReset(email: $email) {
      message
    }
  }
`;

export { CREATE_ITEM, UPDATE_ITEM, DELETE_ITEM, SIGN_UP, SIGN_IN, SIGN_OUT, REQUEST_RESET };
