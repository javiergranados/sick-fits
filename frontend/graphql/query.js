import gql from 'graphql-tag';
import { perPage } from '../config';

const GET_ITEMS = gql`
  query($skip: Int = 0, $first: Int = ${perPage}) {
    items(skip: $skip, first: $first, orderBy: createdAt_DESC) {
      id
      title
      price
      description
      image
      largeImage
    }
  }
`;

const GET_ITEM = gql`
  query($id: ID!) {
    item(where: { id: $id }) {
      id
      title
      description
      price
      largeImage
    }
  }
`;

const PAGINATION = gql`
  {
    itemsConnection {
      aggregate {
        count
      }
    }
  }
`;

const CURRENT_USER = gql`
  {
    me {
      id
      name
      email
      permissions
    }
  }
`;

const GET_ALL_USERS = gql`
  {
    users {
      id
      name
      email
      permissions
    }
  }
`;

export { GET_ITEMS, GET_ITEM, PAGINATION, CURRENT_USER, GET_ALL_USERS };
