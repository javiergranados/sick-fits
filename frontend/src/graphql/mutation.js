import gql from 'graphql-tag';

const CREATE_ITEM = gql`
  mutation createItem($title: String!, $description: String!, $price: Int!, $image: String, $largeImage: String) {
    createItem(title: $title, description: $description, price: $price, image: $image, largeImage: $largeImage) {
      id
    }
  }
`;

const UPDATE_ITEM = gql`
  mutation updateItem($id: ID!, $title: String, $description: String, $price: Int) {
    updateItem(id: $id, title: $title, description: $description, price: $price) {
      id
      title
      description
      price
    }
  }
`;

const DELETE_ITEM = gql`
  mutation deleteItem($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`;

export { CREATE_ITEM, UPDATE_ITEM, DELETE_ITEM };
