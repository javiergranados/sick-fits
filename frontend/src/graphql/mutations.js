import gql from 'graphql-tag';

const CREATE_ITEM = gql`
  mutation createItem($title: String!, $description: String!, $price: Int!, $image: String, $largeImage: String) {
    createItem(title: $title, description: $description, price: $price, image: $image, largeImage: $largeImage) {
      id
    }
  }
`;

export { CREATE_ITEM };
