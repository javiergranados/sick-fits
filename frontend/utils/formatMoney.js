export default function(amount) {
  const options = {
    style: 'currency',
    currency: 'EUR',
    minimumFractionDigits: 2,
  };
  // if its a whole, dollar amount, leave off the .00
  if (amount % 100 === 0) options.minimumFractionDigits = 0;
  const formatter = new Intl.NumberFormat('it', options);
  return formatter.format(amount / 100);
}
// number.toLocaleString('it', { style: 'currency', currency: 'EUR' }) : '0,00 â‚¬';
