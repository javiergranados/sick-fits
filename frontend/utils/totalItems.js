export default function(cart) {
  return cart.reduce((prev, next) => prev + next.quantity, 0);
}
