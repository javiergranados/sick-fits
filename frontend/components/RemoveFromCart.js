import React from 'react';
import { useMutation } from '@apollo/react-hooks';
import PropTypes from 'prop-types';
import { REMOVE_FROM_CART } from '../graphql/mutation';
import { CURRENT_USER } from '../graphql/query';
import * as S from './styles/RemoveFromCart';

const RemoveFromCart = ({ id }) => {
  const update = (cache, { data }) => {
    const { me } = cache.readQuery({ query: CURRENT_USER });
    cache.writeQuery({
      query: CURRENT_USER,
      data: { me: { ...me, cart: me.cart.filter(item => item.id !== data.removeFromCart.id) } },
    });
  };

  const [removeFromCart, { loading }] = useMutation(REMOVE_FROM_CART, {
    update,
    optimisticResponse: { __typename: 'Mutation', removeFromCart: { __typename: 'CartItem', id } },
  });

  const handleClick = () => {
    try {
      removeFromCart({ variables: { id } });
    } catch (err) {
      alert(err.message);
    }
  };

  return (
    <S.RemoveFromCart title="Delete Item" type="button" disabled={loading} onClick={handleClick}>
      &times;
    </S.RemoveFromCart>
  );
};

RemoveFromCart.propTypes = {
  id: PropTypes.string.isRequired,
};

export default RemoveFromCart;
