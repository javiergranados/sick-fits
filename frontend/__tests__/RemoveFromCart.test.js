import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from '@apollo/react-testing';
import { ApolloConsumer } from 'react-apollo';
import { act } from 'react-dom/test-utils';
import RemoveFromCart from '../components/RemoveFromCart';
import { CURRENT_USER } from '../graphql/query';
import { REMOVE_FROM_CART } from '../graphql/mutation';
import { fakeUser, fakeCartItem } from '../utils/testUtils';

global.alert = console.log; //eslint-disable-line

const mocks = [
  {
    request: { query: CURRENT_USER },
    result: {
      data: {
        me: {
          ...fakeUser(),
          cart: [fakeCartItem({ id: 'abc123' })],
        },
      },
    },
  },
  {
    request: { query: REMOVE_FROM_CART, variables: { id: 'abc123' } },
    result: {
      data: {
        removeFromCart: {
          __typename: 'CartItem',
          id: 'abc123',
        },
      },
    },
  },
];

describe('<RemoveFromCart />', () => {
  it('renders and matches snapshot', async () => {
    const wrapper = mount(
      <MockedProvider>
        <RemoveFromCart id="abc123" />
      </MockedProvider>
    );

    expect(toJSON(wrapper.find('button'))).toMatchSnapshot();
  });

  it('removes the item from cart', async () => {
    let apolloClient;
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <ApolloConsumer>
          {client => {
            apolloClient = client;
            return <RemoveFromCart id="abc123" />;
          }}
        </ApolloConsumer>
      </MockedProvider>
    );

    const res = await apolloClient.query({ query: CURRENT_USER });

    expect(res.data.me.cart).toHaveLength(1);
    expect(res.data.me.cart[0].item.price).toBe(5000);

    wrapper.find('button').simulate('click');
    await act(wait);

    const res2 = await apolloClient.query({ query: CURRENT_USER });
    expect(res2.data.me.cart).toHaveLength(0);
  });
});
