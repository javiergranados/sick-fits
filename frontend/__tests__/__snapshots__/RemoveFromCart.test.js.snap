// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RemoveFromCart /> renders and matches snapshot 1`] = `
<button
  className="RemoveFromCart-gh5fkd-0 dwRCct"
  disabled={false}
  onClick={[Function]}
  title="Delete Item"
  type="button"
>
  Ã—
</button>
`;
